crates/junction-api/src/xds/http.rs

```diff
@@ -287,6 +349,283 @@ impl RouteTimeouts {
+                            .with_field_index("actions", i);
+                    }
+                    Ok((
+                        h.descriptor_key.clone(),
+                        RateLimitBucket::Header(HeaderMatch::Present {
+                            name: h.header_name.clone(),
+                            value: true,
+                        }),
+                    ))
+                }
+                xds_ratelimit::action::ActionSpecifier::HeaderValueMatch(h) => {
+                    if h.descriptor_key.is_empty() {
+                        return Err(Error::new_static("descriptor_key should not be empty"))
+                            .with_field("action_specifier")
+                            .with_field_index("actions", i);
+                    }
+                    // NOTE: we only support the first header value match
+                    if let Some(header) = h.headers.first() {
+                        let header_match = HeaderMatch::from_xds(header)?;
+                        Ok((
+                            h.descriptor_key.clone(),
+                            RateLimitBucket::Header(header_match),
+                        ))
+                    } else {
+                        Err(Error::new_static("no header specified"))
+                            .with_field("action_specifier")
+                            .with_field_index("actions", i)
+                    }
+                }
+                xds_ratelimit::action::ActionSpecifier::QueryParameterValueMatch(q) => {
+                    if q.descriptor_key.is_empty() {
+                        return Err(Error::new_static("descriptor_key should not be empty"))
+                            .with_field("action_specifier")
+                            .with_field_index("actions", i);
+                    }
+                    // NOTE: we only support the first header value match
+                    if let Some(query_param) = q.query_parameters.first() {
+                        let query_match = QueryParamMatch::from_xds(query_param)?;
+                        Ok((
+                            q.descriptor_key.clone(),
+                            RateLimitBucket::QueryParam(query_match),
+                        ))
+                    } else {
+                        Err(Error::new_static("no query param specified"))
+                            .with_field("action_specifier")
+                            .with_field_index("actions", i)
+                    }
+                }
+                _ => Err(Error::new_static("unsupported rate limit match"))
+                    .with_field("action_specifier")
+                    .with_field_index("actions", i),
+            })
+            .collect::<Result<HashMap<_, _>, Error>>()
+    }
+
+    fn to_xds(&self) -> xds_ratelimit::action::ActionSpecifier {
+        match self {
+            RateLimitBucket::Header(header) => match header {
+                HeaderMatch::RegularExpression { name, value } => {
```

Comment:
I'm always torn on this, but I feel like moving this into something like `HeaderMatch::to_xds_rate_limit_action` might be nice here. Up to you.